internal/catalogmetadata/types.go

```diff
@@ -0,0 +1,149 @@
+package catalogmetadata
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	bsemver "github.com/blang/semver/v4"
+
+	"github.com/operator-framework/operator-registry/alpha/declcfg"
+	"github.com/operator-framework/operator-registry/alpha/property"
+)
+
+type Schemas interface {
+	Package | Bundle | Channel
+}
+
+type Package struct {
+	declcfg.Package
+}
+
+type Channel struct {
+	declcfg.Channel
+}
+
+type GVK property.GVK
+
+func (g GVK) String() string {
+	return fmt.Sprintf(`group:"%s" version:"%s" kind:"%s"`, g.Group, g.Version, g.Kind)
+}
+
+type GVKRequired property.GVKRequired
+
+func (g GVKRequired) String() string {
+	return fmt.Sprintf(`group:"%s" version:"%s" kind:"%s"`, g.Group, g.Version, g.Kind)
+}
+
+func (g GVKRequired) AsGVK() GVK {
+	return GVK(g)
+}
+
+type Bundle struct {
+	declcfg.Bundle
+	CatalogName string
+	InChannels  []*Channel
+
+	mu sync.RWMutex
+	// these properties are lazy loaded as they are requested
+	propertiesMap map[string]property.Property
+	bundlePackage *property.Package
+	semVersion    *bsemver.Version
+	providedGVKs  []GVK
+	requiredGVKs  []GVKRequired
+}
+
+func (b *Bundle) Version() (*bsemver.Version, error) {
+	if err := b.loadPackage(); err != nil {
+		return nil, err
+	}
+	return b.semVersion, nil
+}
+
+func (b *Bundle) ProvidedGVKs() ([]GVK, error) {
+	if err := b.loadProvidedGVKs(); err != nil {
+		return nil, err
+	}
+	return b.providedGVKs, nil
+}
+
+func (b *Bundle) RequiredGVKs() ([]GVKRequired, error) {
+	if err := b.loadRequiredGVKs(); err != nil {
+		return nil, err
+	}
+	return b.requiredGVKs, nil
+}
+
+func (b *Bundle) loadPackage() error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.bundlePackage == nil {
+		bundlePackage, err := loadFromProps[property.Package](b, property.TypePackage, true)
+		if err != nil {
+			return err
+		}
+		b.bundlePackage = &bundlePackage
+
+		semVer, err := bsemver.Parse(b.bundlePackage.Version)
+		if err != nil {
+			return fmt.Errorf("could not parse semver %q for bundle '%s': %s", b.bundlePackage.Version, b.Name, err)
+		}
+		b.semVersion = &semVer
+	}
```

Comment:
IMO these should be separated out a bit so that on subsequent `Version()` calls we will continue to get an error back when we try to load it when missing. i.e:
1. `Version()` is called for `Bundle`, but version is missing from package info; line 89 returns an error which can be handled.
2. `Version()` is called again; `b.bundlePackage` is no longer `nil` so we skip loading both fields.
3. We return `b.semVersion` on line 60, which is nil, and no error.
This makes it more difficult to handle this case when doing ops like sorting. I think instead we should do:

```suggestion
	if b.bundlePackage == nil {
		bundlePackage, err := loadFromProps[property.Package](b, property.TypePackage, true)
		if err != nil {
			return err
		}
		b.bundlePackage = &bundlePackage
	}
	if b.semVersion == nil {
		semVer, err := bsemver.Parse(b.bundlePackage.Version)
		if err != nil {
			return fmt.Errorf("could not parse semver %q for bundle '%s': %s", b.bundlePackage.Version, b.Name, err)
		}
		b.semVersion = &semVer
	}
```
Since `b.semVersion` will remain nil we can continue giving the relevant error each time it's requested.